pkg/graph/schema.graphql

```diff
@@ -2316,6 +2365,7 @@ type TRBAdminNote {
   noteText: HTML!
   author: UserInfo!
   isArchived: Boolean!
+  categorySpecificData: TRBAdminNoteCategorySpecificData!
```

Comment:
I'm not 100% sure if it's possible, but I would lift this union up to be a union of the whole note type instead of just the categorical data. It adds repetition of common fields in each note, but we have 5 type definitions of categorical data anyway with this approach.

Put another way, if the front end has to check the category of each note regardless, why not flatten the objects and just check `note.__typename` instead?

If I had to pick between empty placeholders for categorical types and some repetition of fields in note data, I would prefer the latter. This allows for expanding general notes with new fields in the future without having to remove the placeholder, creates parity with the separate mutations, and creates sensible types for querying separate note types in the future if needed.

Reply:
I experimented with this a bit; see [this comment](https://github.com/CMSgov/easi-app/pull/2231#issuecomment-1765408993) and [this branch](https://github.com/CMSgov/easi-app/tree/EASI-3360/trb-admin-note-union) (in particular, https://github.com/CMSgov/easi-app/commit/29735cfcff8727f2262550e3478db1704f914751). The problems with having `TRBAdminNote` be a union type are:
1. If the common fields are repeated between the different category-specific types, with a GraphQL interface to enforce that they all have the common fields, that interface breaks `gqlgen`'s model binding.
2. If all the common fields are repeated without any way to enforce that the category-specific types have the same fields, that opens up a lot of room for error if we have to change the common fields or add more in the future; we'd need to make sure we make the same changes across all five category-specific types. I can live with it on the input types because there's only two common fields (`trbRequestId` and `noteText`), but repeating the 9-10 common fields that are returned from queries seems like it introduces a lot of possibilities for inconsistency.
3. We'd have to create a separate new `TRBAdminNoteUnion` type as the union so existing queries don't break; the frontend code in EASI-3361 would then change to querying against that type, and then we'd need to clean up the old type later.

It might be possible to extract the common fields into a separate type, which each of the category-specific types would contain, so there's something like:
```gql
type TRBAdminNoteCommonFields {
  id: UUID!
  trbRequestId: UUID!
  category: TRBAdminNoteCategory!
  noteText: HTML!
  author: UserInfo!
  isArchived: Boolean!
  createdBy: String!
  createdAt: Time!
  modifiedBy: String
  modifiedAt: Time
}

type TRBAdminNoteGeneralRequest {
  commonFields: TRBAdminNoteCommonFields!
}

type TRBAdminNoteInitialRequestForm {
  commonFields: TRBAdminNoteCommonFields!
  appliesToBasicRequestDetails: Boolean!
  appliesToSubjectAreas: Boolean!
  appliesToAttendees: Boolean!
}

# types for the other three categories...

union TRBAdminNoteUnion = TRBAdminNoteGeneralRequest
  | TRBAdminNoteInitialRequestForm
  | TRBAdminNoteSupportingDocuments
  | TRBAdminNoteConsultSession
  | TRBAdminNoteAdviceLetter
```
But I'm not sure if having the nested fields (instead of flattening them out) would cause issues with the BE storage code and/or the FE form code.

In short: I think the current approach works well enough and doesn't cause any significant problems; other alternatives would either cause more problems or would require more work that probably wouldn't add much value even if they didn't cause issues.

Reply:
> but repeating the 9-10 common fields that are returned from queries seems like it introduces a lot of possibilities for inconsistency

Considering this is a return/query type for the FE, I don't see this causing any inconsistency issues in the data. In the worst case scenario, we forget to return a new field in one of the return types and have to add it. I think this could be solved with a comment noting common fields like you've already done. It would be nice to enforce with an interface, but I think the repetition is acceptable since it's a limitation of Apollo/GQL that we're running up against.

I think the more likely and problematic scenario is that we have a need to query by note category at some point in the future. That would mean either adding complexity to the current resolver and storage method to allow filtering, or making separate resolvers and storage methods and creating the new return types you're wanting to avoid creating anyway. Creating the separate resolvers and note types will avoid a category check on the FE, which leads me to believe that would be the preference if we needed that feature.

In other words, we can avoid the union type when querying notes of a specific category if we have separate note types, but if we put that union in a generic note type, the FE will always need to check the category when querying for any kind of note.

> We'd have to create a separate new TRBAdminNoteUnion type as the union so existing queries don't break; the frontend code in [EASI-3361](https://jiraent.cms.gov/browse/EASI-3361) would then change to querying against that type, and then we'd need to clean up the old type later.

This ticket isn't in progress, so I don't see the issue here.

Reply:
> > but repeating the 9-10 common fields that are returned from queries seems like it introduces a lot of possibilities for inconsistency
> 
> Considering this is a return/query type for the FE, I don't see this causing any inconsistency issues in the data. In the worst case scenario, we forget to return a new field in one of the return types and have to add it. I think this could be solved with a comment noting common fields like you've already done. It would be nice to enforce with an interface, but I think the repetition is acceptable since it's a limitation of Apollo/GQL that we're running up against.

Inconsistency when making changes in the future is what I mean, because there wouldn't be any type safety or anything in the schema to keep those fields from getting out of sync, whether that's adding fields, removing fields, changing fields' type.

> I think the more likely and problematic scenario is that we have a need to query by note category at some point in the future. That would mean either adding complexity to the current resolver and storage method to allow filtering, or making separate resolvers and storage methods and creating the new return types you're wanting to avoid creating anyway. Creating the separate resolvers and note types will avoid a category check on the FE, which leads me to believe that would be the preference if we needed that feature.
> 
> In other words, we can avoid the union type when querying notes of a specific category if we have separate note types, but if we put that union in a generic note type, the FE will always need to check the category when querying for any kind of note.

For one thing, that's a pretty specific case of functionality we _might_ add in the future to worry about. For another, I'm not sure how having a separate type for each category would even help with that, unless we added top-level queries for each individual category? 

> > We'd have to create a separate new TRBAdminNoteUnion type as the union so existing queries don't break; the frontend code in [EASI-3361](https://jiraent.cms.gov/browse/EASI-3361) would then change to querying against that type, and then we'd need to clean up the old type later.
> 
> This ticket isn't in progress, so I don't see the issue here.

It's more work we'd have to do in the future. We'd also need to add an additional field to the `TRBRequest` type now, so the frontend ticket could switch to querying the category-specific data when they're ready.

Reply:
My worry above about nested types being an issue isn't really applicable (since a `categorySpecificData` approach introduces nested types anyways), so we could have a type for the common fields. We'd still have the issues with needing to add a new type + new field on `TRBRequest` for the union (and cleaning it up later), though.

Reply:
> Inconsistency when making changes in the future is what I mean, because there wouldn't be any type safety or anything in the schema to keep those fields from getting out of sync, whether that's adding fields, removing fields, changing fields' type.

Yes, but those types would all sit next to each other in the schema file so I don't think keeping them in sync is that big of a concern.

> For one thing, that's a pretty specific case of functionality we might add in the future to worry about. 

For example, if users were viewing the advice letter, I could see a scenario where we're asked to add a feature to view the relevant notes for the advice letter.

> For another, I'm not sure how having a separate type for each category would even help with that, unless we added top-level queries for each individual category?

Top-level queries is what I am suggesting for each note category. This would return the individual note type and there's no need for the FE to validate category specific data. 

> It's more work we'd have to do in the future.

Yes, but I don't see it as a huge lift.

Reply:
Lee, Clay, and I had a group chat about this - we're going with the current approach of having a union type just for the category-specific data.